### ISSUE TITLE
redisv9: incr, ping commands getting EOF constantly after about a day of successful periodic (once in few seconds) calls

### ISSUE THREAD

**Issue by chitturs** [redis/go-redis#2939]
Issue tracker is used for reporting bugs and discussing new features. Please use
[stackoverflow](https://stackoverflow.com) for supporting issues.

Client: GoLang using redisv9
Server: Azure Redis Cache

Client reads messages from Azure service bus and then updates the Redis Cache using Incr(). This is repeated ad infinitum. Everything works for about a day. The calls are made once in few seconds. After about a day, Incr() returns EOF and _never succeeds_. At that point, ping() also returns EOF.

## Expected Behavior

No failures in Incr() or ping().

## Current Behavior

Incr() and ping() fail with EOF.

## Possible Solution

It is unclear what the problem is. Suspicion is a broken connection. The connection should never be idle for the idle timeout to kick in. The server has an idle timeout of 10 minutes.

## Steps to Reproduce

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"os"
	"strings"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/redis/armredis/v3"
	"github.com/redis/go-redis/v9"
)
	op := &redis.Options{Addr: redisHostName + ":" + fmt.Sprint(*getRes.Properties.SSLPort), Username: msiObjectId, Password: accessToken.Token, TLSConfig: &tls.Config{MinVersion: tls.VersionTLS12}}
	client := redis.NewClient(op)

		for _, message := range messages {
			body := message.Body
			redisCacheKey := strings.Trim(string(body), `"`)
			result, err := redisclient.Incr(ctx, redisCacheKey).Result()
			if err != nil {
				if errors.Is(err, context.Canceled) {
					fmt.Println("context was canceled while incrementing cache, return")
					return
				}
				**fmt.Println(fmt.Errorf("redis cache increment failed for key %s, err %v", redisCacheKey, err))**

				// Check if the connection is still alive, https://redis.io/commands/ping/.
				pong, err := redisclient.Ping(ctx).Result()
				fmt.Println(fmt.Errorf("redis cache ping response %s, err %v, timeouts %d", pong, err, redisclient.PoolStats().Timeouts))
				continue
			}

			fmt.Println("Key:", redisCacheKey, "Result:", result)
		}

## Context (Environment)

My AKS cluster app is partially broken in functionality. The Redis Cache is used for stats purposes of other activities in the cluster. The stats is now not working after about a day.

**Comment by chriscasola** [comment]
We are seeing similar behavior. In our case we have a status check on our service that does a `redisClient.Set(ctx, "status-key", 1, 0)` and eventually that starts returning EOF. This is also effecting our other redis calls once the EOF is encountered, but things seem to recover on their own without a restart of the service.

I'm not sure how to debug this, but I also suspect a connection in the pool that is in a bad state.

Should it be possible for a connection in the pool to be unusable?

**Comment by ccbhj** [comment]
What's your go-redis version? maybe you coud try v9.4.0.  It works for me but not v9.5.x or v9.6.x

**Comment by ad-zsolt-imre** [comment]
Both go-redis `v9.4.0` and `v9.6.1` cannot establish a TLS connection to a perfectly functioning Redis server. I tested the server by connecting using `redis-cli` and TLS enabled and was working perfectly. But, no matter what I do, the go client won't work. In the Redis logs I see the connection coming in (see below), and then it keeps repeating multiple times until finally go-redis gives up and reports the `EOF`. The client connection count always stays `0`.

```
1:M 10 Aug 2024 14:16:34.738 - Accepted 10.6.97.104:38212
1:M 10 Aug 2024 14:16:36.703 - DB 0: 1 keys (0 volatile) in 4 slots HT.
1:M 10 Aug 2024 14:16:36.703 . 0 clients connected (0 replicas), 1242704 bytes in use
```

**Comment by bzssm** [comment]
because the access token is only valid for 1 day?

**Comment by ccbhj** [comment]
> because the access token is only valid for 1 day?

No. After a connection is established go-redis will send a command [CLIENT SETINFO](https://redis.io/docs/latest/commands/client-setinfo/) but Some redis client implementation(for example codis) will close connections to warn you that such a command are disabled or not supported, which causes EOF for whatever commands your send after connections established

**Comment by ccbhj** [comment]
Anyway the main reason is that your redis server somehow close the connection after established. Maybe you could check why your redis servers close it.

**Comment by ndyakov** [comment]
Doesn't look like client issue, closing this for now. Feel free to reopen if needed.
